### PR TITLE
Crash loop fix and other changes

### LINKED ISSUE #181: Concurrent calls to GetEPGForChannel - crash loop on start LibreElec
On startup concurrent calls can occur to GetEPGForChannel. Confirmed on LibreElec for for x86 and Rpi.

Crashlog below shows the call occurring on multiple threads.

Fix is to use a mutex on this call in Epg class to ensure one calll at a time.

### PR DESCRIPTION
**Pull request by phunkyfish**
v3.17.1
- Fixed: Ensure only one call to GetEPGForChannel happens at one time, fixes #181
- Added: Update/fix server version for pvr addon
- Added: Option to skip initial EPG Load
- Fixed: Fix for zap on channel change for dreamboxes
- Added: Support for padding in certain timer types
- Fixed: Change the call web/tunersignal to web/signal so it also works on DreamOS

Crashlog shows 2 calls on two different thread to GetEPGForChannel on LibreElec, reproduced on both x86 and Rpi.

### CODE REVIEW

**Comment by phunkyfish** [comment]
@ksooo a nice small PR. Not sure why this only happens on LibreElec. I think it's been around a while just happens occasionally.

**Comment by phunkyfish** [comment]
So the failed fix brings us from crash loop to a deadlock and only on LibreElec (before and after the fix). I'll need to play this one a bit to figure it out.

**Comment by MilhouseVH** [comment]
I can't think of any specific differences in the threading model used by LibreELEC that would cause this issue. As is often the case with these things it may just be down to timing, or other startup add-on activity in LibreELEC means it is more likely to trigger. I'd put money on the same issue being present in (for instance) Ubuntu, but it may simply be harder to reproduce.

File: src/Enigma2.cpp [hunk 1]
```diff
@@ -295,7 +295,15 @@ PVR_ERROR Enigma2::GetEPGForChannel(ADDON_HANDLE handle, const PVR_CHANNEL &chan
...
     Sleep(m_settings.GetEPGDelayPerChannelDelay());
 
-  return m_epg.GetEPGForChannel(handle, channel, iStart, iEnd);
+  if (!m_channels.IsValid(channel.iUniqueId))
+  {
+    Logger::Log(LEVEL_ERROR, "%s Could not fetch channel object - not fetching EPG for channel with UniqueID '%d'", __FUNCTION__, channel.iUniqueId);
+    return PVR_ERROR_SERVER_ERROR;
+  }
+
+  std::shared_ptr<Channel> myChannel = m_channels.GetChannel(channel.iUniqueId);
```
Comment (ksooo) [inline src/Enigma2.cpp]: const std::shared_ptr

File: src/Enigma2.cpp [hunk 2]
```diff
@@ -295,7 +295,15 @@ PVR_ERROR Enigma2::GetEPGForChannel(ADDON_HANDLE handle, const PVR_CHANNEL &chan
...
+
+  return m_epg.GetEPGForChannel(handle, myChannel, iStart, iEnd);
```
Comment (ksooo) [inline src/Enigma2.cpp]: Why not making one line out of the two above? For what do you need myChannel in a var?

File: src/enigma2/Epg.cpp [hunk 3]
```diff
@@ -42,15 +45,10 @@ void Epg::TriggerEpgUpdatesForChannels()
   }
 }
 
-PVR_ERROR Epg::GetEPGForChannel(ADDON_HANDLE handle, const PVR_CHANNEL &channel, time_t iStart, time_t iEnd)
+PVR_ERROR Epg::GetEPGForChannel(ADDON_HANDLE handle, std::shared_ptr<Channel> &myChannel, time_t iStart, time_t iEnd)
```
Comment (ksooo) [inline src/enigma2/Epg.cpp]: const std::shared_ptr... &

File: src/enigma2/Epg.cpp [hunk 4]
```diff
@@ -42,15 +45,10 @@ void Epg::TriggerEpgUpdatesForChannels()
...
 {
-  if (!m_channels.IsValid(channel.iUniqueId))
-  {
-    Logger::Log(LEVEL_ERROR, "%s Could not fetch channel object - not fetching EPG for channel with UniqueID '%d'", __FUNCTION__, channel.iUniqueId);
-    return PVR_ERROR_NO_ERROR;
-  }
-
-  std::shared_ptr<Channel> myChannel = m_channels.GetChannel(channel.iUniqueId);
+  //Let's ensure we only run one EPG update at a time
+  //CLockObject lock(m_mutex);
```
Comment (ksooo) [inline src/enigma2/Epg.cpp]: I would remove that line, not comment it.

File: src/enigma2/Epg.h [hunk 5]
```diff
@@ -31,18 +31,21 @@
...
   {
   public:
-    Epg (enigma2::Channels &channels, enigma2::ChannelGroups &channelGroups, enigma2::extract::EpgEntryExtractor &entryExtractor);
+    Epg(enigma2::Channels &channels, enigma2::ChannelGroups &channelGroups, enigma2::extract::EpgEntryExtractor &entryExtractor);
+    Epg(const enigma2::Epg& epg);
 
     bool IsInitialEpgCompleted();
     void TriggerEpgUpdatesForChannels();
-    PVR_ERROR GetEPGForChannel(ADDON_HANDLE handle, const PVR_CHANNEL &channel, time_t iStart, time_t iEnd);
+    PVR_ERROR GetEPGForChannel(ADDON_HANDLE handle, std::shared_ptr<enigma2::data::Channel> &channel, time_t iStart, time_t iEnd);
```
Comment (ksooo) [inline src/enigma2/Epg.h]: const std::shared_ptr... &

**Comment by ksooo** [comment]
> I can't think of any specific differences in the threading model used by LibreELEC that would cause this issue.

Agreed. The problem is platform agnostic but for some reason LE seems to trigger this more often than other platforms.

**Comment by phunkyfish** [comment]
The way the addon currently works it first loads now and next EPG data followed by full EPG all based off the GetEPGForChannel call. I could try to load the now and next beforehand so it’s simply a read when GetEPGForChannel is called. 

Maybe as a first step I’ll disable the now and next and see if that makes a difference. I can repro it consistently now I just need to use the testbuilds as I can’t build libreElec myself.

@ksooo the PR currently is a hack so I can try to find a fix it’s by no means finished and will have several interations while I figure it out. Will make sure it’s all cleaned up once I get it sorted.

**Comment by phunkyfish** [comment]
@ksooo I figure the latest will fix the issue. I moved the initial EPG load to before (also much quicker now), so it can be a simple read if called. So no contention across groups which caused the previous problem.

**Comment by phunkyfish** [comment]
For last commit I missed the call to GetServerVersion. Will push the change shortly.

**Comment by phunkyfish** [comment]
Ok, good to go

**Comment by phunkyfish** [comment]
Fix confirmed in testbuilds, so we should be good.

**Comment by phunkyfish** [comment]
There's a bug in the update server version commit. Will update it shortly.

**Comment by phunkyfish** [comment]
@ksooo this is finally ready to go! A not so small PR after all ;)

**Comment by phunkyfish** [comment]
Another crash loop unfortunately, will try skipping initial EPG load.

**Comment by phunkyfish** [comment]
So I added a default option to skip the initial EPG load (now and next) which successfully avoids the Crash Loop issue. Will revisit and look for cleaner solution when I work on the Async Connection feature.

**Comment by phunkyfish** [comment]
So @ksooo this is ready to go when you are ;)

**Comment by phunkyfish** [comment]
There is a bug in the padding commit. Will post back once fixed.

**Comment by phunkyfish** [comment]
@ksooo, all fixes tested by users in testbuilds. Ready for your review when you have the time.

File: src/enigma2/Epg.h [hunk 6]
```diff
@@ -21,43 +21,54 @@
...
-
-    bool GetInitialEPGForGroup(std::shared_ptr<enigma2::data::ChannelGroup> group);
-    PVR_ERROR GetInitialEPGForChannel(ADDON_HANDLE handle, const std::shared_ptr<enigma2::data::Channel> &channel, time_t iStart, time_t iEnd);
+    PVR_ERROR TransferInitialEPGForChannel(ADDON_HANDLE handle, const std::shared_ptr<data::EpgChannel> &epgChannel, time_t iStart, time_t iEnd);
+    std::shared_ptr<data::EpgChannel> GetEpgChannel(const std::string serviceReference);
+    bool LoadInitialEPGForGroup(const std::shared_ptr<enigma2::data::ChannelGroup> group);
+    bool ChannelNeedsInitialEpg(const std::string serviceReference);
+    bool InitialEpgLoadedForChannel(const std::string serviceReference);
+    bool InitialEpgReadForChannel(const std::string serviceReference);
+    std::shared_ptr<data::EpgChannel> GetEpgChannelNeedingInitialEpg(const std::string serviceReference);
```
Comment (ksooo) [inline src/enigma2/Epg.h]: ref vs. copy for almost all new private member functions...

File: src/enigma2/data/EpgChannel.h [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    public:
+
+      EpgChannel() = default;
+      EpgChannel(const EpgChannel &e) : BaseChannel(e) {};
+      ~EpgChannel() = default;
+
+      bool RequiresInitialEpg() const { return m_requiresInitialEpg; }
+      void SetRequiresInitialEpg(bool value) { m_requiresInitialEpg = value; }
+
+      std::vector<EpgEntry>& GetInitialEPG() { return m_initialEPG; }
```
Comment (ksooo) [inline src/enigma2/data/EpgChannel.h]: This member function could be const.

File: src/enigma2/data/EpgEntry.cpp [hunk 8]
```diff
@@ -37,33 +37,35 @@ void EpgEntry::UpdateTo(EPG_TAG &left) const
   left.iFlags              = EPG_TAG_FLAG_UNDEFINED;
 }
 
-bool EpgEntry::UpdateFrom(TiXmlElement* eventNode, Channels &channels)
+bool EpgEntry::UpdateFrom(TiXmlElement* eventNode, std::map<std::string, std::shared_ptr<EpgChannel>> &m_epgChannelsMap)
```
Comment (ksooo) [inline src/enigma2/data/EpgEntry.cpp]: m_ prefix should only be used for member vars.

File: src/enigma2/data/EpgEntry.cpp [hunk 9]
```diff
@@ -37,33 +37,35 @@ void EpgEntry::UpdateTo(EPG_TAG &left) const
...
 {
-  std::string strTmp;
+  std::string m_serviceReference;
```
Comment (ksooo) [inline src/enigma2/data/EpgEntry.cpp]: same here. m_ prefix should only be used for member vars.

File: src/enigma2/Admin.cpp [hunk 10]
```diff
@@ -19,6 +19,13 @@ using namespace enigma2::extract;
 using namespace enigma2::utilities;
 using json = nlohmann::json;
 
+Admin::Admin() : m_addonVersion(STR(VUPLUS_VERSION))
+{
+  std::string unknown = "unknown";
```
Comment (ksooo) [inline src/enigma2/Admin.cpp]: Is it inteded to have this string not localized?

File: pvr.vuplus/changelog.txt [hunk 11]
```diff
@@ -1,6 +1,7 @@
 v3.17.1
 - Fixed: Ensure only one call to GetEPGForChannel happens at one time, fixes #181
 - Added: Update/fix server version for pvr addon
+- Added: Option to skip initial EPG Load
```
Comment (ksooo) [inline pvr.vuplus/changelog.txt]: Imo, functional extensions justify a minor version bump. micro bumps are for fixes only.

File: src/enigma2/Epg.cpp [hunk 12]
```diff
@@ -154,6 +154,18 @@ void Epg::TriggerEpgUpdatesForChannels()
   }
 }
 
+void Epg::MarkChannelAsInitialEpgRead(const std::string serviceReference)
```
Comment (ksooo) [inline src/enigma2/Epg.cpp]: ref or copy?

File: src/enigma2/Epg.cpp [hunk 13]
```diff
@@ -154,6 +154,18 @@ void Epg::TriggerEpgUpdatesForChannels()
...
+
+  if (epgChannel->RequiresInitialEpg())
+  {
+    epgChannel->SetRequiresInitialEpg(false);
+    epgChannel->GetInitialEPG().clear();
+    m_readInitialEpgChannelsMap.erase(epgChannel->GetServiceReference());
+  }
+}
+
 PVR_ERROR Epg::GetEPGForChannel(ADDON_HANDLE handle, const std::string serviceReference, time_t iStart, time_t iEnd)
```
Comment (ksooo) [inline src/enigma2/Epg.cpp]: not added by this PR, but: ref vs. copy

File: pvr.vuplus/addon.xml.in [hunk 14]
```diff
@@ -180,6 +180,7 @@ v3.17.1
 - Added: Update/fix server version for pvr addon
 - Added: Option to skip initial EPG Load
 - Fixed: Fix for zap on channel change for dreamboxes
+- Added: Support for padding in certain timer types
```
Comment (ksooo) [inline pvr.vuplus/addon.xml.in]: this definitely justifies a minor version bump

File: src/enigma2/data/Timer.cpp [hunk 15]
```diff
@@ -260,12 +274,47 @@ bool Timer::UpdateFrom(TiXmlElement* timerNode, Channels &channels)
     }
   }
 
+  if (ContainsTag(TAG_FOR_PADDING))
+  {
+    std::string paddingTag = ReadTag(TAG_FOR_PADDING);
```
Comment (ksooo) [inline src/enigma2/data/Timer.cpp]: const std::string? Di you need the var at all? Its value is used only one time.

File: src/Enigma2.cpp [hunk 16]
```diff
@@ -156,7 +159,8 @@ void *Enigma2::Process()
       Sleep(5 * 1000);
   }
 
-  // Trigger "Real" EPG updates
+  // Whether or not initial EPG updates occurred now Trigger "Real" EPG updates
+  // This will reagrd Initial EPG as completed anyway.
```
[the remark this answers is not in the document]
Comment (phunkyfish) [reply src/Enigma2.cpp]: Fixed

File: src/enigma2/Epg.cpp [hunk 17]
```diff
@@ -15,229 +17,242 @@ using namespace enigma2;
 using namespace enigma2::data;
 using namespace enigma2::extract;
 using namespace enigma2::utilities;
+using namespace P8PLATFORM;
 using json = nlohmann::json;
 
-Epg::Epg (enigma2::Channels &channels, enigma2::ChannelGroups &channelGroups, enigma2::extract::EpgEntryExtractor &entryExtractor)
-      : m_channels(channels), m_channelGroups(channelGroups), m_entryExtractor(entryExtractor) {}
+Epg::Epg(enigma2::extract::EpgEntryExtractor &entryExtractor)
```
[the remark this answers is not in the document]
Comment (phunkyfish) [reply src/enigma2/Epg.cpp]: Yes it is

File: src/enigma2/Epg.cpp [hunk 18]
```diff
@@ -15,229 +17,242 @@ using namespace enigma2;
...
-  for (auto& channel : m_channels.GetChannelsList())
-  {
-    Logger::Log(LEVEL_DEBUG, "%s - Trigger EPG update for channel '%d'", __FUNCTION__, channel->GetUniqueId());
-    PVR->TriggerEpgUpdate(channel->GetUniqueId());
-  }
-}
+Epg::Epg(const Epg &epg) : m_entryExtractor(epg.m_entryExtractor) {}
 
-PVR_ERROR Epg::GetEPGForChannel(ADDON_HANDLE handle, const PVR_CHANNEL &channel, time_t iStart, time_t iEnd)
+bool Epg::Initialise(enigma2::Channels &channels, enigma2::ChannelGroups &channelGroups)
```
[the remark this answers is not in the document]
Comment (phunkyfish) [reply src/enigma2/Epg.cpp]: The classes return vectors so can't be const.

File: src/enigma2/Epg.cpp [hunk 19]
```diff
@@ -15,229 +17,242 @@ using namespace enigma2;
...
-    if (!m_allChannelsHaveInitialEPG)
-      m_allChannelsHaveInitialEPG = m_channels.CheckIfAllChannelsHaveInitialEPG();
+    newEpgChannel.SetRadio(channel->IsRadio());
+    newEpgChannel.SetUniqueId(channel->GetUniqueId());
+    newEpgChannel.SetChannelName(channel->GetChannelName());
+    newEpgChannel.SetServiceReference(channel->GetServiceReference());
 
-    if (m_allChannelsHaveInitialEPG)
-      Logger::Log(LEVEL_DEBUG, "%s - Intial EPG update COMPLETE!", __FUNCTION__);
+    m_epgChannels.emplace_back(new EpgChannel(newEpgChannel));
```
[the remark this answers is not in the document]
Comment (phunkyfish) [reply src/enigma2/Epg.cpp]: Well it's obvious now :(

[the remark this answers is not in the document]
Comment (phunkyfish) [reply src/enigma2/Epg.cpp]: Done

File: src/enigma2/Epg.cpp [hunk 20]
```diff
@@ -15,229 +17,242 @@ using namespace enigma2;
...
 
-    return GetInitialEPGForChannel(handle, myChannel, iStart, iEnd);
+    m_epgChannelsMap.insert({epgChannel->GetServiceReference(), epgChannel});
+    m_readInitialEpgChannelsMap.insert({epgChannel->GetServiceReference(), epgChannel});
+    m_needsInitialEpgChannelsMap.insert({epgChannel->GetServiceReference(), epgChannel});
   }
 
-  const std::string url = StringUtils::Format("%s%s%s",  Settings::GetInstance().GetConnectionURL().c_str(), "web/epgservice?sRef=",
-                                              WebUtils::URLEncodeInline(myChannel->GetServiceReference()).c_str());
+  int lastScannedIgnoreSuccessCount = round((1 - LAST_SCANNED_INITIAL_EPG_SUCCESS_PERCENT) * m_epgChannels.size());
```
[the remark this answers is not in the document]
Comment (phunkyfish) [reply src/enigma2/Epg.cpp]: I blame google

File: src/enigma2/Epg.cpp [hunk 21]
```diff
@@ -15,229 +17,242 @@ using namespace enigma2;
...
-
-  TiXmlDocument xmlDoc;
-  if (!xmlDoc.Parse(strXML.c_str()))
+  std::vector<std::shared_ptr<ChannelGroup>> groupList;
+  ChannelGroup newChannelGroup;
+  newChannelGroup.SetRadio(false);
+  newChannelGroup.SetGroupName("Last Scanned"); //Name not important
+  newChannelGroup.SetServiceReference("1:7:1:0:0:0:0:0:0:0:FROM BOUQUET \"userbouquet.LastScanned.tv\" ORDER BY bouquet");
+  newChannelGroup.SetLastScannedGroup(true);
+  groupList.emplace_back(new ChannelGroup(newChannelGroup));
```
[the remark this answers is not in the document]
Comment (phunkyfish) [reply src/enigma2/Epg.cpp]: Done

File: src/enigma2/Epg.cpp [hunk 22]
```diff
@@ -15,229 +17,242 @@ using namespace enigma2;
...
 
-  for (; pNode != nullptr; pNode = pNode->NextSiblingElement("e2event"))
-  {
-    EpgEntry entry;
+  return true;
+}
 
-    if (!entry.UpdateFrom(pNode, myChannel, iStart, iEnd))
-      continue;
+std::shared_ptr<data::EpgChannel> Epg::GetEpgChannel(const std::string serviceReference)
```
[the remark this answers is not in the document]
Comment (phunkyfish) [reply src/enigma2/Epg.cpp]: Sorry, it's just silly

[the remark this answers is not in the document]
Comment (phunkyfish) [reply src/enigma2/Epg.cpp]: Done

File: src/enigma2/Epg.h (hunk 6 above)
Comment (phunkyfish) [reply src/enigma2/Epg.h]: Done

File: src/enigma2/data/EpgChannel.h (hunk 7 above)
Comment (phunkyfish) [reply src/enigma2/data/EpgChannel.h]: It returns a vector, which would be required to be const for this to be possible? Or am I missing something.

File: src/enigma2/data/EpgEntry.cpp (hunk 8 above)
Comment (phunkyfish) [reply src/enigma2/data/EpgEntry.cpp]: Done

File: src/enigma2/data/EpgEntry.cpp (hunk 9 above)
Comment (phunkyfish) [reply src/enigma2/data/EpgEntry.cpp]: Done

File: src/enigma2/Admin.cpp (hunk 10 above)
Comment (phunkyfish) [reply src/enigma2/Admin.cpp]: Sorry, this will be overwritten, but should just be null terminated.

File: src/enigma2/Epg.cpp (hunk 13 above)
Comment (phunkyfish) [reply src/enigma2/Epg.cpp]: Done

File: pvr.vuplus/addon.xml.in (hunk 14 above)
Comment (phunkyfish) [reply pvr.vuplus/addon.xml.in]: Done

File: src/enigma2/data/Timer.cpp (hunk 15 above)
Comment (phunkyfish) [reply src/enigma2/data/Timer.cpp]: Done

**Comment by phunkyfish** [comment]
@ksooo I couldn't cleanly fixup the EPG changes so added a cleanup commit. Other than that we should be good.

**Comment by phunkyfish** [comment]
Reminder ping!

File: src/enigma2/Epg.cpp [hunk 23]
```diff
@@ -40,30 +40,31 @@ bool Epg::Initialise(enigma2::Channels &channels, enigma2::ChannelGroups &channe
...
 
   std::vector<std::shared_ptr<ChannelGroup>> groupList;
-  ChannelGroup newChannelGroup;
-  newChannelGroup.SetRadio(false);
-  newChannelGroup.SetGroupName("Last Scanned"); //Name not important
-  newChannelGroup.SetServiceReference("1:7:1:0:0:0:0:0:0:0:FROM BOUQUET \"userbouquet.LastScanned.tv\" ORDER BY bouquet");
-  newChannelGroup.SetLastScannedGroup(true);
-  groupList.emplace_back(new ChannelGroup(newChannelGroup));
+
+  std::shared_ptr<ChannelGroup> newChannelGroup = std::make_shared<data::ChannelGroup>();
```
Comment (ksooo) [inline src/enigma2/Epg.cpp]: If you only want to call the default ctor, std::make_shared does not make to much sense. A simple `std::shared_ptr<ChannelGroup> newChannelGroup;` does exactly the same.

File: src/enigma2/Epg.cpp [hunk 24]
```diff
@@ -40,30 +40,31 @@ bool Epg::Initialise(enigma2::Channels &channels, enigma2::ChannelGroups &channe
   //add an initial EPG data per channel uId, sref and initial EPG
   for (auto& channel : channels.GetChannelsList())
   {
-    EpgChannel newEpgChannel;
+    std::shared_ptr<data::EpgChannel> newEpgChannel = std::make_shared<data::EpgChannel>();
```
Comment (ksooo) [inline src/enigma2/Epg.cpp]: If you only want to call the default ctor, std::make_shared does not make to much sense. A simple `std::shared_ptr<EpgChannel> newEpgChannel;` does exactly the same.

File: src/enigma2/Epg.cpp (hunk 23 above)
Comment (phunkyfish) [reply src/enigma2/Epg.cpp]: Obvious now that you say it. Fixed

File: src/enigma2/Epg.cpp (hunk 24 above)
Comment (phunkyfish) [reply src/enigma2/Epg.cpp]: And again

**Comment by phunkyfish** [comment]
Ty ;)

**Pull request merged**